Tutorial 5 - 3-D Topography

weiji14 opened the pull request:
Tutorial showcasing 3D perspective plots over Mars and Antarctica using `grdview`.

Preview image for Olympus Mons, Mars:

![3D perspective plot of Olympus Mons, Mars](https://github.com/user-attachments/assets/b751043b-c132-4cce-839d-4caaec01b3ed)

Preview image for Antarctica:

![3D perspective plot of Ross Island, Antarctica](https://github.com/user-attachments/assets/2bc45b25-f3b0-44db-9f23-5add3c4112bc)

Notes:
- If anyone wants to test this locally, clone this branch and run `jupyter-book build book/`
- I'll convert the *.md file (done in MyST Markdown) into a Jupyter Notebook before merging this PR. Using a markdown file for now just to make it easier to review the code.

TODO:
- [x] Show how to use `grdview` for elevation plot with simple colormap over Mars
- [x] Show how to use `grdview` to create 3D plot with draped imagery over Antarctica
- [ ] etc

References:
- https://github.com/andrebelem/PlanetaryMaps/blob/main/2_%20Tutorial%20Mars%20Maps%20V2.ipynb
- https://github.com/GenericMappingTools/egu22pygmt/pull/5

Fixes #11

weiji14 committed:
3D topography tutorial part 1 - Mars

Tutorial showcasing Mars Orbiter Laser Altimeter (MOLA) dataset, focusing on Olympus Mons. Data loaded via `pygmt.datasets.load_mars_relief`, plotting in 2D with `grdimage`, and in 3D using `grdview`.

 data processing pipeline over Wellington, New Zealand. Doing data loading using laspy, pre-processing with `blockmedian`, gridding with `surface`, and plotting in 2D with `grdimage` and in 3D using `grdview`.

weiji14 committed:
Add tut5_topography to TOC and Neumann et al., 2003 citation

Include tut5_topography in Table of Contents, and citation for the MOLA dataset.

weiji14 committed:
3D topography part 2a - Antarctica

Draft section on plotting an Antarctic DEM with RGB image overlay. Using REMA as the DEM, and Sentinel-2 L2A imagery. Need to mosaic two tiles together using rioxarray for both the DEM and RGB imagery. Have done a quick `grdview` plot based on my old code at https://github.com/weiji14/30DayMapChallenge2021/commit/61f5698c742b642e0b28b175c9a8ebfb63da2321.

andrebelem commented:
@weiji14  
I cloned the repository and tested it on my local machine. It seems that `rioxarray` needs to be included in the `environment.yml` file. Is this correct, or am I missing something?

andrebelem commented:
![image](https://github.com/user-attachments/assets/eba4f9d0-3b9f-45bb-92ca-5a2fa9e4e8fc)

weiji14 commented on environment.yml:
```diff
@@ -10,3 +10,4 @@ dependencies:
   # Optional dependencies
   - geopandas=1.0.1
   - jupyter-book=1.0.2
+  - rioxarray=0.17.0
```
> I cloned the repository and tested it on my local machine. It seems that rioxarray needs to be included in the environment.yml file. Is this correct, or am I missing something?

Yes, added here already :smile:

andrebelem replied:
Hello @weiji14,

I’ve tested everything here, and I was planning to add a few more explanations, especially about the pros and cons of data resolution and computational requirements for large mosaics. Should I make these changes directly in the tut5_topography.md file, or should I create a new branch?

weiji14 replied:
Yes, feel free to push changes directly (be sure to do a `git pull` first in case I push any changes too :wink:). I'll be adding more text to the Antarctica section too later.

andrebelem committed:
A few explanations in grdview

yvonnefroehlich commented on book/tut5_topography.md:
```diff
@@ -0,0 +1,258 @@
+centered at approximately 19°N and 133°W, with a height of 22 km (14 miles) and
+approximately 700 km (435 miles) in diameter.
+
+Let's grab a higher resolution map over the volcano, and plot another 2D map!
```
```suggestion
Let's grab a higher resolution map over the volcano, and plot another 2-D map!
```

yvonnefroehlich commented on book/tut5_topography.md:
````diff
@@ -0,0 +1,258 @@
+da_mars
+```
+
+## 2D map view
````
```suggestion
## 2-D map view
```

yvonnefroehlich commented on book/tut5_topography.md:
```diff
@@ -0,0 +1,258 @@
+
+# 3D Topography 🏔️
+
+In this tutorial, let's use PyGMT to create 3D perspective plots of Digital Elevation
```
```suggestion
In this tutorial, let's use PyGMT to create 3-D perspective plots of Digital Elevation
```

yvonnefroehlich commented on book/tut5_topography.md:
```diff
@@ -0,0 +1,258 @@
+  name: python3
+---
+
+# 3D Topography 🏔️
```
Maybe use a hypen here to be consistent with the PyGMT documentation?

```suggestion
# 3-D Topography 🏔️
```

yvonnefroehlich commented on book/tut5_topography.md:
````diff
@@ -0,0 +1,258 @@
+fig.show()
+```
+
+# 1️⃣ Using `grdview` for 3D Visualization
````
```suggestion
# 1️⃣ Using `grdview` for 3-D Visualization
```

yvonnefroehlich commented on book/tut5_topography.md:
```diff
@@ -0,0 +1,258 @@
+# 1️⃣ Using `grdview` for 3D Visualization
+
+The [`grdview`](https://www.pygmt.org/v0.13.0/api/generated/pygmt.Figure.grdview.html)
+function in PyGMT is a powerful tool for creating 3D perspective views of gridded data.
```
```suggestion
method in PyGMT is a powerful tool for creating 3-D perspective views of gridded data.
```

yvonnefroehlich commented on book/tut5_topography.md:
```diff
@@ -0,0 +1,258 @@
+  with azimuth can help showcase specific aspects of the data from different angles.
+- **Elevation** (`elevation`): Controls the vertical angle of the view, with 90°
+  representing a top-down perspective and lower values providing more of a side view.
+  Typically, values between 20° and 60° create a balanced 3D effect.
```
```suggestion
  Typically, values between 20° and 60° create a balanced 3-D effect.
```

yvonnefroehlich commented on book/tut5_topography.md:
```diff
@@ -0,0 +1,258 @@
+- **zsize** - vertical exaggeration as z-axis size, we used `4c` here for 4 centimeters
+- **surftype** - using `s` will just create a regular surface
+- **shading** - set to `True` to get the default hillshading effect
+- **frame** - A proper 3D map frame that consists of:
```
```suggestion
- **frame** - A proper 3-D map frame that consists of:
```

yvonnefroehlich commented on book/tut5_topography.md:
````diff
@@ -0,0 +1,258 @@
+When working with DEM mosaics and optical imagery, carefully consider the size and resolution of the data. High-resolution DEMs combined with complex topographies can demand substantial computational resources for processing and visualization. A practical tip is to start with lower resolutions to experiment with and refine the scene geometry (e.g., azimuth, elevation, and perspective). Once you a ...
+```
+
+# 3️⃣ Draping RGB image on 3D topography
````
```suggestion
# 3️⃣ Draping RGB image on 3-D topography
```

yvonnefroehlich commented on book/tut5_topography.md:
````diff
@@ -0,0 +1,258 @@
+fig.show()
+```
+
+When setting the `zscale` for vertical exaggeration, choose a value that balances clarity and realism. For subtle topographies, higher exaggeration (e.g., smaller `zscale` values) can emphasize elevation differences, making features more visible. However, for steep terrains, lower exaggeration helps maintain a natural appearance. You may use `shading=True` to add hillshading, which enhances the 3 ...
````
```suggestion
When setting the `zscale` for vertical exaggeration, choose a value that balances clarity and realism. For subtle topographies, higher exaggeration (e.g., smaller `zscale` values) can emphasize elevation differences, making features more visible. However, for steep terrains, lower exaggeration helps maintain a natural appearance. You may use `shading=True` to add hillshading, which enhances the 3-D effect by simulating light and shadows, making terrain features easier to interpret. Experiment with both parameters to find the best combination for your dataset and visualization goals.
```

yvonnefroehlich commented on book/tut5_topography.md:
```diff
@@ -0,0 +1,258 @@
+- **surftype** - using `s` will just create a regular surface
+- **shading** - set to `True` to get the default hillshading effect
+- **frame** - A proper 3D map frame that consists of:
+  - automatic tick marks on x and y axis (e.g. `xaf` and `yaf`)
```
```suggestion
  - automatic tick marks on x and y axis (e.g., `xaf` and `yaf`)
```

yvonnefroehlich commented on book/tut5_topography.md:
````diff
@@ -0,0 +1,258 @@
+
+```{code-cell}
+fig = pygmt.Figure()
+fig.grdimage(grid=da_mars, frame=True, projection="W12c", cmap="SCM/managua+h0")
````
```suggestion
fig.grdimage(grid=da_mars, frame=True, projection="W12c", cmap="managua+h0")
```

yvonnefroehlich commented on book/tut5_topography.md:
````diff
@@ -0,0 +1,258 @@
+
+```{code-cell}
+fig = pygmt.Figure()
+fig.grdimage(grid=da_olympus, frame=["WSne+tOlympus Mons", "af"], cmap="SCM/managua+h0")
````
```suggestion
fig.grdimage(grid=da_olympus, frame=["WSne+tOlympus Mons", "af"], cmap="managua+h0")
```

yvonnefroehlich commented on book/tut5_topography.md:
```diff
@@ -0,0 +1,258 @@
+
+fig.grdview(
+    grid=da_olympus,
+    cmap="SCM/managua+h0",
```
```suggestion
    cmap="managua+h0",
```

yvonnefroehlich commented on book/tut5_topography.md:
```diff
@@ -0,0 +1,258 @@
+[Mollweide projection](https://www.pygmt.org/v0.13.0/projections/misc/misc_mollweide.html).
+To get a reddish hue, we'll use a
+[colormap](https://docs.generic-mapping-tools.org/6.5/reference/cpts.html)
+called `SCM/managua` which comes with a soft hinge
```
Hm. For me, it works to directly give the name of a Scientific colourmap (without the folder "SCM"). We are also doing this in the PyGMT examples.

```suggestion
called `managua` which comes with a soft hinge
```

weiji14 replied:
Yes, there's backwards compatibility with just using `managua` vs `SCM/managua` (see https://github.com/GenericMappingTools/gmt/pull/6446#issuecomment-1126711471). The `SCM/` prefix is intended for better recognition that this is a Scientific Colour Map, see comment at https://docs.generic-mapping-tools.org/6.5/reference/cpts.html:

> The color maps are subdivided into a number of sections relating to their source. This avoids name clashes and improves recognision of the color maps as well as their authors

yvonnefroehlich replied:
Thanks for pointing out this GMT PR! Hm, maybe we should then do this for all tutorials consistently.

yvonnefroehlich reviewed (commented):
These maps look quite cool and visually appealing! I especially like the Antarctica map 🙂 !

> I'll convert the *.md file (done in MyST Markdown) into a Jupyter Notebook before merging this PR. Using a markdown file for now just to make it easier to review the code.

Yeah, reviewing code and docs in plain text, i.e., normal Python scripts or at least markdown, is so much more convenient!

> If anyone wants to test this locally, clone this branch and run jupyter-book build book/

Will try this, hopefully tomorrow, for a more detailed review.

weiji14 committed:
3D to 3-D

Co-authored-by: Yvonne Fröhlich <94163266+yvonnefroehlich@users.noreply.github.com>

weiji14 committed:
Update book/tut5_topography.md

Co-authored-by: Yvonne Fröhlich <94163266+yvonnefroehlich@users.noreply.github.com>

weiji14 committed:
Add descriptive text for 'Draping RGB image on 3-D topography' section

Explain some of the parameters used for image draping such as `drapegrid`, `surftype` and `zscale`. Moved some explanatory text into a 'tip' admonition cell.

weiji14 committed:
WIP on add-jn-topo

weiji14 committed:
Rename from tut5_topography to tut05_topography

weiji14 committed:
Wrap some tips to 88 characters

weiji14 committed:
Revert accidental changes

Patch ad34d2900cb4e1b95f70cd6fc8767aefd1c0ecf0

weiji14 committed:
Add links to other 3D perspective plot examples

weiji14 commented:
@andrebelem or anyone who wants to review, let me know if there's anything else to include, otherwise I'll merge this tomorrow!

andrebelem commented:
> @andrebelem or anyone who wants to review, let me know if there's anything else to include, otherwise I'll merge this tomorrow!

Hey @weiji14,

It looks great to me! Yes, please go ahead and merge it as planned.

weiji14 committed:
Merge branch 'main' into add-jn-topo

weiji14 committed:
Convert MyST markdown notebook to Jupyter notebook

Using `jupytext --to notebook book/tut05_topography.md`

weiji14 committed:
Reduce diff
